client/signup/config/flows.js

```diff
@@ -238,11 +252,82 @@ const Flows = {
 			flow = filterDesignTypeInFlow( flow );
 		}
 
-		return flow;
+		if ( ! user.get() ) {
+			// Preload the AB tests before going forward in the flow.
+			// Only valid for logged out users.
+			Flows.preloadABTestVariationsForStep( flowName, currentStepName );
+		}
+
+		return Flows.getABTestFilteredFlow( flowName, flow );
 	},
 
 	getFlows() {
 		return flows;
+	},
+
+	/**
+	 * Preload AB Test variations after a certain step has been completed.
+	 *
+	 * This gives the option to set the AB variation as late as possible in the
+	 * signup flow.
+	 *
+	 * Currently only the `main` flow is whitelisted.
+	 *
+	 * @param {String} flowName The current flow
+	 * @param {String} stepName The step that is being completed right now
+	 */
+	preloadABTestVariationsForStep( flowName, stepName ) {
+
+		/**
+		 * In cases where the flow is being resumed, the flow must not be changed from what the user
+		 * has seen before.
+		 *
+		 * E.g. A user is resuming signup from before the test was added. There is no need
+		 * to add a step somewhere back in the line.
+		 */
+		if ( Flows.resumingFlow ) {
+			return;
+		}
+
+		if ( 'main' === flowName ) {
+			if ( 'survey' === stepName ) {
+				abtest( 'siteTitleStep' );
+			}
+		}
+	},
+
+	/**
+	 * Return a flow that is modified according to the ABTest rules.
+	 *
+	 * Useful when testing new steps in the signup flows.
+	 *
+	 * Example usage: Inject or remove a step in the flow if a user is part of an ABTest.
+	 *
+	 * @param {String} flowName The current flow name
+	 * @param {Object} flow The flow object
+	 */
+	getABTestFilteredFlow( flowName, flow ){
+		const updatedFlow = Object.assign( {}, flow );
```

Comment:
> Major: assign performs shallowly, meaning that flow.steps === updatedFlow.steps even though flow !== updatedFlow. 😱

😥 (disappointed but relieved face - such a proper description). Thanks for catching that, today I learned! :) Updated it with `lodash`'s `cloneDeep` to make sure it's properly cloned.

> Minor: this assignment is performed even when flowName !== 'main', due to how the function is structured.

It definitely makes more of a performance sense to put the expensive clone inside the whitelist check, to make sure there aren't any extra calls to it.